allowing keyword arguments to pass to json3.read from oxygen.json

shikon opened the pull request:
In order to parse json files which included inf and nan's it is required to pass along the keyword arguments.
oxygen.json is not able to do this yet, therefore adding this functionality.

I am leaving open the combination of custom struct + passing keyword arguments of the form json(request, classtype; keywordargs), this can be added later if required.

ndortega commented:
@shikon Thanks for the PR, great catch! Could you add that additional use case to this pr? If we're going to make these changes to this module, we should try to cover all the use cases. If you can't, I can add this use case + tests when I get some time this week.

shikon commented:
> @shikon Thanks for the PR, great catch! Could you add that additional use case to this pr? If we're going to make these changes to this module, we should try to cover all the use cases. If you can't, I can add this use case + tests when I get some time this week.

Sure I will do that later this week. Lets keep the PR open for now.

shikon commented:
I will put a few extra commits with the additional use cases, but please double check if the tests pass. I am not able to use your run test file succesfully on my local computer and I don't see any CI/CD here on github.

shikon commented:
![image](https://user-images.githubusercontent.com/33011458/235096176-1384bf8d-8802-42b0-9942-f5005564cdd6.png)
Already in your build pipeline there are errors... Maybe just close this PR, if you have time later to show me how I can make all test pass then I can continue I already have something in mind.

ndortega commented:
There are errors that show up during the tests, but these are expected and shouldn't stop the tests from running. Aside from that, I've moved your test case into its own file. I moved it out because your tests don't require a running server in order to work. 

1. I'd recommend cd'ing into the `/test` directory and opening a new julia repl there directly. 
2. activate & instantiate 
3. include("bodyparsertests.jl")
4. This will let you quickly run your specific tests in isolation from the other tests

shikon commented:
> There are errors that show up during the tests, but these are expected and shouldn't stop the tests from running. Aside from that, I've moved your test case into its own file. I moved it out because your tests don't require a running server in order to work.
> 
>     1. I'd recommend cd'ing into the `/test` directory and opening a new julia repl there directly.
> 
>     2. activate & instantiate
> 
>     3. include("bodyparsertests.jl")
> 
>     4. This will let you quickly run your specific tests in isolation from the other tests

Hi,
Thanks for the steps, I am experienced developing in julia ;).
The tests that I had in mind would require to be added the current file as they are going to use internalrequest and are a response.
I do need a little help still understanding what errors are intended, how can I judge that?

ndortega commented:
The easiest way to know if an error is expected is if the test is looking for a 500 response code. Otherwise, the test would be wrapped in a `try...catch` block.

That being said if it's possible to execute your test case without adding it to the server in the `runtests.jl` file, I would prefer that. This file is already pretty massive, and I would rather move any new tests to smaller modules. Since `json()` is a pure function, any test that can be performed using `internalrequest()` with the router and server can also be performed without it.

shikon commented:
I thought you had some preference making the test in that way.
Then I will keep on doing it like I did.

shikon commented:
I think its ready to be merged.

regarding local testing of runtest:
this is one of the errors I get
![image](https://user-images.githubusercontent.com/33011458/235440124-6f52d24f-8078-459a-a177-30c6993ab806.png)

shikon reviewed (commented):
Looks good to me, indeed redundant json response is created when you can put it one function.
Awesome!
